RecoParticleFlow/PFClusterProducer/plugins/PFClusterFromHGCalMultiCluster.cc

```diff
@@ -14,43 +17,69 @@ void PFClusterFromHGCalMultiCluster::buildClusters(const edm::Handle<reco::PFRec
   const auto& hgcalMultiClusters = *clusterH_;
   auto const& hits = *input;
 
+  const auto& tracksters = *trackstersH_;
+
   // for quick indexing back to hit energy
   std::unordered_map<uint32_t, size_t> detIdToIndex(hits.size());
   for (uint32_t i = 0; i < hits.size(); ++i) {
     detIdToIndex[hits[i].detId()] = i;
   }
 
+  int iMultiClus = -1;
+
   for (const auto& mcl : hgcalMultiClusters) {
+    iMultiClus++;
+
+    // Filter using trackster PID
+    if (filterByTracksterPID_) {
+      float probTotal = 0.0f;
+
+      for (int& cat : filter_on_categories_) {
+        probTotal += tracksters[iMultiClus].id_probabilities(cat);
+      }
+
+      if (probTotal < pid_threshold_) {
+        continue;
+      }
+    }
+
     DetId seed;
     double energy = 0.0, highest_energy = 0.0;
     output.emplace_back();
     reco::PFCluster& back = output.back();
-    for (const auto& cl : mcl) {
-      const auto& hitsAndFractions = cl->hitsAndFractions();
-      for (const auto& hAndF : hitsAndFractions) {
-        auto itr = detIdToIndex.find(hAndF.first);
-        if (itr == detIdToIndex.end()) {
-          continue;  // hit wasn't saved in reco
-        }
-        auto ref = makeRefhit(input, itr->second);
-        assert(ref->detId() == hAndF.first.rawId());
-        const double hit_energy = hAndF.second * ref->energy();
-        energy += hit_energy;
-        back.addRecHitFraction(reco::PFRecHitFraction(ref, hAndF.second));
-        // TODO: the following logic to identify the seed of a cluster
-        // could be appropriate for the Run2 Ecal Calorimetric
-        // detector, but could be wrong for the HGCal one. This has to
-        // be reviewd.
-        if (hit_energy > highest_energy || highest_energy == 0.0) {
-          highest_energy = hit_energy;
-          seed = ref->detId();
-        }
-      }  // end of hitsAndFractions
-    }    // end of loop over clusters (2D/layer)
-    if (energy <= 1) {
```

Comment:
should this condition be retained in the new algorithm?

Reply:
@kpedro88 This is certainly something that needs to be studied further, but it seems to be working reasonably well for now as far as e/gamma reconstruction is concerned.